File: go/88.merge-sorted-array.go
/*
 * @lc app=leetcode id=88 lang=golang
 *
 * [88] Merge Sorted Array
 */

// @lc code=start
func merge(nums1 []int, m int, nums2 []int, n int) {
	r1 := m - 1
	r2 := n - 1
	for r := m + n - 1; r >= 0 && r1 >= 0 && r2 >= 0; r-- {
		if nums1[r1] > nums2[r2] {
			nums1[r] = nums1[r1]
			r1--
		} else {
			nums1[r] = nums2[r2]
			r2--
		}
	}
	for r2 >= 0 {
		nums1[r2] = nums2[r2]
		r2--
	}
}

// @lc code=end

File: java/88.merge-sorted-array.java
/*
 * @lc app=leetcode id=88 lang=java
 *
 * [88] Merge Sorted Array
 */
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        if (nums1 == null || nums2 == null) return;
        int i = m-1, j = n-1, k = m+n-1;
        while (k > i) {
            if (i == -1) {
                for (;k >= 0;) nums1[k--] = nums2[j--];
                break;
            }
            nums1[k--] = nums1[i] < nums2[j] ? nums2[j--] : nums1[i--];
        }
    }
}

